charts/k8s-service/templates/canarydeployment.yaml

```diff
@@ -0,0 +1,325 @@
+
+Similarly, we need to decide whether or not there are environment variables to add
+
+We need this because certain sections are omitted if there are no volumes or environment variables to add.
+*/ -}}
+
+{{/* Go Templates do not support variable updating, so we simulate it using dictionaries */}}
+{{- if .Values.canary.enabled -}}
+{{- $hasInjectionTypes := dict "hasVolume" false "hasEnvVars" false "hasIRSA" false "exposePorts" false -}}
+{{- if .Values.envVars -}}
+  {{- $_ := set $hasInjectionTypes "hasEnvVars" true -}}
+{{- end -}}
+{{- $allContainerPorts := values .Values.containerPorts -}}
+{{- range $allContainerPorts -}}
+  {{/* We are exposing ports if there is at least one key in containerPorts that is not disabled (disabled = false or
+       omitted)
+  */}}
+  {{- if or (not (hasKey . "disabled")) (not .disabled) -}}
+    {{- $_ := set $hasInjectionTypes "exposePorts" true -}}
+  {{- end -}}
+{{- end -}}
+{{- if gt (len .Values.aws.irsa.role_arn) 0 -}}
+  {{- $_ := set $hasInjectionTypes "hasEnvVars" true -}}
+  {{- $_ := set $hasInjectionTypes "hasVolume" true -}}
+  {{- $_ := set $hasInjectionTypes "hasIRSA" true -}}
+{{- end -}}
+{{- $allSecrets := values .Values.secrets -}}
+{{- range $allSecrets -}}
+  {{- if eq (index . "as") "volume" -}}
+    {{- $_ := set $hasInjectionTypes "hasVolume" true -}}
+  {{- else if eq (index . "as") "environment" -}}
+    {{- $_ := set $hasInjectionTypes "hasEnvVars" true -}}
+  {{- else if eq (index . "as") "none" -}}
+    {{- /* noop */ -}}
+  {{- else -}}
+    {{- fail printf "secrets config has unknown type: %s" (index . "as") -}}
+  {{- end -}}
+{{- end -}}
+{{- $allConfigMaps := values .Values.configMaps -}}
+{{- range $allConfigMaps -}}
+  {{- if eq (index . "as") "volume" -}}
+    {{- $_ := set $hasInjectionTypes "hasVolume" true -}}
+  {{- else if eq (index . "as") "environment" -}}
+    {{- $_ := set $hasInjectionTypes "hasEnvVars" true -}}
+  {{- else if eq (index . "as") "none" -}}
+    {{- /* noop */ -}}
+  {{- else -}}
+    {{- fail printf "configMaps config has unknown type: %s" (index . "as") -}}
+  {{- end -}}
+{{- end -}}
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: {{ include "k8s-service.fullname" . }}-canary
+  labels:
+    # These labels are required by helm. You can read more about required labels in the chart best practices guide:
+    # https://docs.helm.sh/chart_best_practices/#standard-labels
+    helm.sh/chart: {{ include "k8s-service.chart" . }}
+    app.kubernetes.io/name: {{ include "k8s-service.name" . }}
+    app.kubernetes.io/instance: {{ .Release.Name }}
+    app.kubernetes.io/managed-by: {{ .Release.Service }}
+{{- with .Values.deploymentAnnotations }}
+  annotations:
+{{ toYaml . | indent 4 }}
+{{- end }}
+spec:
+  replicas: {{ .Values.canary.replicaCount | default 1 }}
+  selector:
+    matchLabels:
+      app.kubernetes.io/name: {{ include "k8s-service.name" . }}
+      app.kubernetes.io/instance: {{ .Release.Name }}
```

Comment:
Since this selector is shared with the main deployment, both Deployment controllers will now manage pods from BOTH controllers. You need to make sure there is another label so that it doesn't overlap.

Reply:
Ah gotcha. Initially I thought this was desired such that the service would correctly route traffic to both deployments, but that totally makes sense. Will fix.

Reply:
The selectors for the `Service` are defined on the `Service`: https://github.com/gruntwork-io/helm-kubernetes-services/blob/master/charts/k8s-service/templates/service.yaml#L30. These selectors are for indicating which pods the `Deployment` controller should manage.

So you want the `Deployment` selectors to uniquely identify the pods, but the `Service` selectors to overlap.

Reply:
After reviewing and playing around with it locally, I believe this should be resolved in ac924b0

Reply:
I think the proposed fix has the problem where now the canary pods won't be pulled into the `Service`.

Does it not work if you add a third label to both Deployments, one indicating canary one that is not? E.g:

```
# canary deployment
gruntwork.io/deployment-type: canary

# main deployment
gruntwork.io/deployment-type: main
```

In theory, this should work because you keep the original selectors so they get caught in the Service, but now the pod selectors are unique between the two deployments.

Reply:
Makes sense - let me try this

Reply:
I think that looks good: 

```bash
# Service selector - gets both deployments 
$ k get pods -l app.kubernetes.io/name=wikka-wakka,app.kubernetes.io/instance=musty-stoat
NAME                                              READY   STATUS    RESTARTS   AGE
musty-stoat-wikka-wakka-6dcf8b5b7f-pt8cs          1/1     Running   0          12m
musty-stoat-wikka-wakka-canary-6c4546d8f4-bz8hv   1/1     Running   0          12m

# Main deployment selector - only gets main 
$ k get pods -l app.kubernetes.io/name=wikka-wakka,app.kubernetes.io/instance=musty-stoat,gruntwork.io/deployment-type=main
NAME                                       READY   STATUS    RESTARTS   AGE
musty-stoat-wikka-wakka-6dcf8b5b7f-pt8cs   1/1     Running   0          12m

# Canary deployment selector - only gets canary
$ k get pods -l app.kubernetes.io/name=wikka-wakka,app.kubernetes.io/instance=musty-stoat,gruntwork.io/deployment-type=canary
NAME                                              READY   STATUS    RESTARTS   AGE
musty-stoat-wikka-wakka-canary-6c4546d8f4-bz8hv   1/1     Running   0          12m
```

Reply:
Fixed in cdfc453

Reply:
nice!